src/Features/Core/Portable/AliasAmbiguousType/AbstractAliasAmbiguousTypeCodeFixProvider.cs

```diff
@@ -0,0 +1,91 @@
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using static Microsoft.CodeAnalysis.CodeActions.CodeAction;
+
+namespace Microsoft.CodeAnalysis.AliasAmbiguousType
+{
+    internal abstract class AbstractAliasAmbiguousTypeCodeFixProvider : CodeFixProvider
+    {
+
+        protected abstract string GetTextPreviewOfChange(string aliasName, ITypeSymbol typeSymbol);
+
+        public override FixAllProvider GetFixAllProvider() => null;
+
+        public sealed override async Task RegisterCodeFixesAsync(CodeFixContext context)
+        {
+            var cancellationToken = context.CancellationToken;
+            var document = context.Document;
+            var syntaxFacts = document.GetLanguageService<ISyntaxFactsService>();
+            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+
+            // Innermost: We are looking for an IdentifierName. IdentifierName is sometimes at the same span as its parent (e.g. SimpleBaseTypeSyntax).
+            var diagnosticNode = root.FindNode(context.Span, getInnermostNodeForTie: true);
+            if (!syntaxFacts.IsIdentifierName(diagnosticNode))
+            {
+                return;
+            }
+
+            var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+            var symbolInfo = semanticModel.GetSymbolInfo(diagnosticNode, cancellationToken);
+            if (SymbolCandidatesContainsSupportedSymbols(symbolInfo))
+            {
+                var addImportService = document.GetLanguageService<IAddImportsService>();
+                var syntaxGenerator = document.GetLanguageService<SyntaxGenerator>();
+                var compilation = semanticModel.Compilation;
+                var optionSet = await document.GetOptionsAsync(cancellationToken).ConfigureAwait(false);
+                var placeSystemNamespaceFirst = optionSet.GetOption(GenerationOptions.PlaceSystemNamespaceFirst, document.Project.Language);
+                var codeActionsBuilder = ImmutableArray.CreateBuilder<CodeAction>(symbolInfo.CandidateSymbols.Length);
+                foreach (var symbol in symbolInfo.CandidateSymbols.Cast<ITypeSymbol>())
+                {
+                    var typeName = symbol.Name;
+                    var codeActionPreviewText = GetTextPreviewOfChange(typeName, symbol);
+                    Task<Document> CreateChangedDocument(CancellationToken c)
+                    {
+                        var aliasDirective = syntaxGenerator.AliasImportDeclaration(typeName, symbol);
+                        var newRoot = addImportService.AddImport(compilation, root, diagnosticNode, aliasDirective, placeSystemNamespaceFirst);
+                        return Task.FromResult(document.WithSyntaxRoot(newRoot));
+                    };
+                    codeActionsBuilder.Add(new MyCodeAction(codeActionPreviewText, CreateChangedDocument));
```

Comment:
nit: our pattern is to generally use a lambda here.  but it's not a big deal.  if you prefer the above form, then keep it.